File: 351004/Leshok/lab3/publisher/src/services/StickerService.js
const { Sticker, News } = require('../models');
const { StickerResponseDto } = require('../dtos/StickerDto');

class StickerService {
  async create(dto) {
    const { name } = dto;
    const sticker = await Sticker.create({ name });
    return new StickerResponseDto(sticker.id, sticker.name);
  }

  async update(id, dto) {
    const sticker = await Sticker.findByPk(id);
    if (!sticker) throw { status: 404, code: '40403', message: 'Sticker not found' };
    const { name } = dto;
    await sticker.update({ name });
    return new StickerResponseDto(sticker.id, sticker.name);
  }

  async getById(id) {
    const sticker = await Sticker.findByPk(id);
    if (!sticker) throw { status: 404, code: '40403', message: 'Sticker not found' };
    return new StickerResponseDto(sticker.id, sticker.name);
  }

  async getAll() {
    const stickers = await Sticker.findAll();
    return stickers.map(s => new StickerResponseDto(s.id, s.name));
  }

  async delete(id) {
    const sticker = await Sticker.findByPk(id);
    if (!sticker) throw { status: 404, code: '40403', message: 'Sticker not found' };
    await sticker.destroy();
  }
}

module.exports = new StickerService();

File: 351004/Beliaev/restapp/publisher/src/main/java/com/example/publisher/service/StickerService.java
package com.example.publisher.service;

import com.example.publisher.dto.request.StickerRequestTo;
import com.example.publisher.dto.response.StickerResponseTo;
import com.example.publisher.exception.EntityNotFoundException;
import com.example.publisher.mapper.StickerMapper;
import com.example.publisher.model.Sticker;
import com.example.publisher.repository.StickerRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.cache.annotation.CacheEvict;
import org.springframework.cache.annotation.CachePut;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.cache.annotation.Caching;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.stream.Collectors;

@Service
@Transactional(readOnly = true)
@RequiredArgsConstructor
public class StickerService {
    private final StickerRepository repository;
    private final StickerMapper mapper;

    @Transactional
    @Caching(
            put = @CachePut(value = "sticker", key = "#result.id"),
            evict = @CacheEvict(value = "stickers_list", allEntries = true)
    )
    public StickerResponseTo create(StickerRequestTo request) {
        Sticker sticker = mapper.toEntity(request);
        Sticker saved = repository.saveAndFlush(sticker);
        return mapper.toResponse(saved);
    }

    @Cacheable(value = "stickers_list")
    public List<StickerResponseTo> getAll() {
        return repository.findAll().stream()
                .map(mapper::toResponse)
                .collect(Collectors.toList());
    }

    @Cacheable(value = "sticker", key = "#id")
    public StickerResponseTo getById(Long id) {
        return repository.findById(id)
                .map(mapper::toResponse)
                .orElseThrow(() -> new EntityNotFoundException("Sticker not found with id: " + id));
    }

    @Transactional
    @Caching(
            put = @CachePut(value = "sticker", key = "#id"),
            evict = @CacheEvict(value = "stickers_list", allEntries = true)
    )
    public StickerResponseTo update(Long id, StickerRequestTo request) {
        Sticker sticker = repository.findById(id)
                .orElseThrow(() -> new EntityNotFoundException("Sticker not found with id: " + id));
        mapper.updateEntityFromDto(request, sticker);
        Sticker saved = repository.save(sticker);
        return mapper.toResponse(saved);
    }

    @Transactional
    @Caching(evict = {
            @CacheEvict(value = "sticker", key = "#id"),
            @CacheEvict(value = "stickers_list", allEntries = true)
    })
    public void delete(Long id) {
        if (!repository.existsById(id)) {
            throw new EntityNotFoundException("Sticker not found with id: " + id);
        }
        repository.deleteById(id);
    }
}